Weakness: Add fuzzy matching, expand dex search to support types

bingingem opened the pull request:
https://www.smogon.com/forums/threads/support-fuzzy-searching-in-weak.3731963/

This is based on work done by @earljoves in #9929 - I expanded it to include type matching

earljoves commented:
Sorry, been busy the past couple of weeks! Appreciate the continued work, @bingingem!

bingingem commented:
I thought better of renaming the dex function, since it was out of scope of this change. I rebased this branch and removed that extra change, so the only thing in this PR should just be to expand /weak to allow for fuzzy matching, including fuzzy matching type names.
